### PR TITLE
Guide for OpenAPI generation

### PR DESCRIPTION
**Pull request by GuentherJulian**
PR for https://github.com/devonfw/devon4j/issues/483
Added a guide to OpenAPI describing what OpenAPI is and extensions/plugins to automatically create OpenAPI specifications from REST APIs. For the Maven plugin ServicedocGen, I've linked to the Quarkus reference application as we plan to modify the project to use the plugin in the future.

### CODE REVIEW

File: documentation/guide-openapi.asciidoc [hunk 1]
```diff
@@ -0,0 +1,94 @@
+:toc: macro
+toc::[]
+
+= OpenAPI
+
+The link:https://spec.openapis.org/oas/latest.html[OpenAPI Specification (OAS)] defines a standard for describing RESTful web services in a machine- and human-readable format. OpenAPI allows REST APIs to be defined in a uniform manner.
+Technically, an OpenAPI document is written in YAML or JSON format. The specification defines the structure of a REST API by describing attributes such as path information, response codes, and return types. Some examples can be found link:https://github.com/OAI/OpenAPI-Specification/tree/main/examples/v3.0[here].
+
+OpenAPI is often used in combination with link:https://swagger.io/[Swagger]. Swagger is a set of tools build around OpenAPI, that help developers to design and document their REST APIs.
+The most common tool is the link:https://swagger.io/tools/swagger-ui/[Swagger UI], which uses the OpenAPI specification to create a graphical interface of the REST API that you can also interact with. Check out the link:https://editor.swagger.io/[Swagger online editor] to get a feeling of this.
```
Comment (mustafaokyay) [inline documentation/guide-openapi.asciidoc:10]: Check out the link:https://editor.swagger.io/[Swagger online editor] to get a feeling of (->for) this (->it).

File: documentation/guide-openapi.asciidoc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+[source,xml]
+--------
+<dependency>
+  <groupId>org.springdoc</groupId>
+  <artifactId>springdoc-openapi-ui</artifactId>
+  <version>1.5.12</version>
+</dependency>
+--------
+
+The extension analysis the REST API classes at runtime and then generates the documentation from them. The documentation can be extended by using link:https://github.com/swagger-api/swagger-core/wiki/Annotations-1.5.X[Swagger API annotations].
```
Comment (mustafaokyay) [inline documentation/guide-openapi.asciidoc:27]: the extension analysis(-> analyzes) ...

File: documentation/guide-openapi.asciidoc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+The OpenAPI specification can be accessed by requesting `/q/openapi`.
+
+Smallrye OpenAPI is compliant with link:https://github.com/eclipse/microprofile-open-api[MicroProfile OpenAPI]. You can add MicroProfile annotations to further describe your REST endpoints and extend the OpenAPI documentation.
+More information for this can be found link:https://quarkus.io/blog/openapi-for-everyone/#openapi[here] or link:https://download.eclipse.org/microprofile/microprofile-open-api-1.0/microprofile-openapi-spec.html#_documentation_mechanisms[here]. 
+
+== ServicedocGen Maven Plugin
+
+The link:https://github.com/mojohaus/servicedocgen-maven-plugin[ServicedocGen maven plugin] can be used within both Spring and Quarkus applications.
+It works a bit different then the extensions mentioned above. The plugin analysis the REST API and it's JavaDoc and then generate the OpenAPI specification and the Swagger UI as static files. So no Swagger or MicroProfile annotations have to be added.
```
Comment (mustafaokyay) [inline documentation/guide-openapi.asciidoc:55]: The plugin analysis(-> analyzes) ...

**Review by maybeec — CHANGES_REQUESTED** [review]
Some high-level comments on it.
1. I like that you showed the options a project might have, but options also come with the burden of taking a decision. That's what devonfw wants to get rid of for stupid things or at least give clear guidance on the decision. So, we should provide a short comparison at the start enabling people to make their decision easily. Additionally, we should claim, that if there is no clarity of the decision to take choice X - on quarkus this might be smallrye, on spring that might be servicedocgen.
2. Why don't we provide a short devonfw-tutorials/tutorials tutorial for each of the options. It#s a quite short tutorial, which anyhow easily shows how it works and also shows hands-on how the swagger UI looks like for the cases where it can be generated. Also the people can play around and try additional cases out by hand to understand the boundaries of the tool. In this guide then, we should simply reference the different options to be tried out in katacoda.

File: documentation/guide-openapi.asciidoc (hunk 2 above)
Comment (GuentherJulian) [reply documentation/guide-openapi.asciidoc:27]: I think both spellings are correct, don't you?

**Comment by GuentherJulian** [comment]
> Some high-level comments on it.
> 
> 1. I like that you showed the options a project might have, but options also come with the burden of taking a decision. That's what devonfw wants to get rid of for stupid things or at least give clear guidance on the decision. So, we should provide a short comparison at the start enabling people to make their decision easily. Additionally, we should claim, that if there is no clarity of the decision to take choice X - on quarkus this might be smallrye, on spring that might be servicedocgen.
> 2. Why don't we provide a short devonfw-tutorials/tutorials tutorial for each of the options. It#s a quite short tutorial, which anyhow easily shows how it works and also shows hands-on how the swagger UI looks like for the cases where it can be generated. Also the people can play around and try additional cases out by hand to understand the boundaries of the tool. In this guide then, we should simply reference the different options to be tried out in katacoda.

Totally agree with you that we should give clear guidance for decision on which extension/plugin to choose. I will do some rework on it.
The hint with the tutorials is also a good one. I created an issue for that https://github.com/devonfw-tutorials/tutorials/issues/179

File: documentation/guide-openapi.asciidoc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+Technically, an OpenAPI document is written in YAML or JSON format. The specification defines the structure of a REST API by describing attributes such as path information, response codes, and return types. Some examples can be found link:https://github.com/OAI/OpenAPI-Specification/tree/main/examples/v3.0[here].
+
+OpenAPI is often used in combination with link:https://swagger.io/[Swagger]. Swagger is a set of tools build around OpenAPI, that help developers to design and document their REST APIs.
+The most common tool is the link:https://swagger.io/tools/swagger-ui/[Swagger UI], which uses the OpenAPI specification to create a graphical interface of the REST API that you can also interact with. Check out the link:https://editor.swagger.io/[Swagger online editor] to get a feeling for it.
+
+== OpenAPI generation
+
+There are several extensions you can use in your project to automatically generate the OpenAPI specifications and Swagger UI from your REST API (code-first approach). devon4j recommends the following two extensions/plugins to use:
+
+* Smallrye OpenAPI extension (for Quarkus)
```
Comment (maybeec) [inline documentation/guide-openapi.asciidoc:16]: Do we?
I thought we recommend ServiceDocGen also for Quarkus if you don't want to re-specify all rest service information again as annotations. ServiceDocGen should be also very interesting for the Quarkus stack in this case when implementing JAX-RS standard.

Comment (maybeec) [reply documentation/guide-openapi.asciidoc:16]: Would be great to say, Quarkus recommends Smallrye whereas devonfw recommends strongly to have a look at ServiceDocGen if you want to reduce effort on REST service documentation by omitting additional documentation annotations needed for smallrye. Nonetheless, in smallrye, you can document services in much more detail if needed.

**Review by maybeec — CHANGES_REQUESTED** [review]

File: documentation/guide-openapi.asciidoc (hunk 4 above)
Comment (GuentherJulian) [reply documentation/guide-openapi.asciidoc]: Did some rework on it

**Review by maybeec — APPROVED** [review]

**Pull request merged**